test/Microsoft.Azure.Devices.Edge.Test.Common/TomlDocument.cs

```diff
@@ -24,67 +22,64 @@ public void ReplaceOrAdd<T>(string dottedKey, T value)
             {
                 string tableKey = segments[i];
 
-                try
+                if (!table.ContainsKey(tableKey))
                 {
-                    table = (TomlTable)table[tableKey];
-                }
-                catch (KeyNotFoundException)
-                {
-                    // Nett does not provide a function to easily add table subkeys.
-                    // A hack workaround is to serialize the table, append the new table as a string, then deserialize.
-                    // This only needs to be done once to create the new subtable.
-                    // After that, Nett functions can be used to modify the subtable.
-                    string tableName = string.Join(".", segments.Take(segments.Length - 1));
-                    this.AddTable(tableName, segments[segments.Length - 1], value);
-                    return;
+                    if (add)
+                    {
+                        table.Add(tableKey, table.CreateEmptyAttachedTable());
+                    }
+                    else
+                    {
+                        return (table, string.Empty);
+                    }
                 }
+
+                table = (TomlTable)table[tableKey];
             }
 
-            string key = segments[segments.Length - 1];
+            return (table, segments[segments.Length - 1]);
```

Comment:
Why are we dropping the last bit?

Reply:
The "dotted key" that gets passed to `TraverseKey` is composed of zero or more dot-separated section names, followed by a dot, and finally the key name (e.g., "sectionA.sectionB.key"). `TraverseKey` parses `dottedKey` into an array of strings, where the first n-1 elements are section names (e.g., ["sectionA","sectionB"]) and the last element is the key (e.g., "key").

This return statement isn't dropping the last bit, it is returning the innermost section object and the key. So if you pass in "sectionA.sectionB.key", you'll get back a tuple that is the sectionB TomlTable object and the "key" string.